File: jester-api/src/main/java/io/quartic/jester/api/DatasetLocator.java
package io.quartic.jester.api;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property="type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = PostgresDatasetLocator.class, name = "postgres"),
        @JsonSubTypes.Type(value = GeoJsonDatasetLocator.class, name = "geojson"),
        @JsonSubTypes.Type(value = WebsocketDatasetLocator.class, name = "websocket"),
})
public interface DatasetLocator {
}

File: catalogue-api/src/main/kotlin/io/quartic/catalogue/api/model/DatasetLocator.kt
package io.quartic.catalogue.api.model

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeInfo.As
import com.fasterxml.jackson.annotation.JsonTypeInfo.Id
import io.quartic.catalogue.api.model.DatasetLocator.*

@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type")
@JsonSubTypes(
        Type(value = PostgresDatasetLocator::class, name = "postgres"),
        Type(value = GeoJsonDatasetLocator::class, name = "geojson"),
        Type(value = WebsocketDatasetLocator::class, name = "websocket"),
        Type(value = CloudDatasetLocator::class, name = "cloud"),
        Type(value = GooglePubSubDatasetLocator::class, name = "google-pubsub")
)
interface DatasetLocator {
    data class PostgresDatasetLocator(
            val user: String,
            val password: String,
            val url: String,
            val query: String
    ) : DatasetLocator

    data class GeoJsonDatasetLocator(val url: String) : DatasetLocator

    data class WebsocketDatasetLocator(val url: String) : DatasetLocator

    data class GooglePubSubDatasetLocator(val topic: String) : DatasetLocator

    data class CloudDatasetLocator(
            val path: String,
            val streaming: Boolean = false,
            val mimeType: String
    ) : DatasetLocator
}
